crates/wasmtime/src/compile/call_graph.rs

```diff
@@ -0,0 +1,103 @@
+//! Construction of the call-graph.
+
+#![cfg_attr(not(test), expect(dead_code, reason = "used in upcoming PRs"))]
+
+use super::*;
+use core::{
+    fmt::{self, Debug},
+    ops::Range,
+};
+use wasmtime_environ::{EntityRef, SecondaryMap};
+
+/// A call graph reified into a densely packed and quickly accessible
+/// representation.
+///
+/// In a call graph, nodes are functions, and an edge `f --> g` means that the
+/// function `f` calls the function `g`.
+pub struct CallGraph<Node>
+where
+    Node: EntityRef + Debug,
+{
+    /// A map from each node to the subslice of `self.edge_elems` that are its
+    /// edges.
+    edges: SecondaryMap<Node, Range<u32>>,
+
+    /// Densely packed edge elements for `self.edges`.
+    edge_elems: Vec<Node>,
+}
+
+impl<Node> Debug for CallGraph<Node>
+where
+    Node: EntityRef + Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        struct Edges<'a, Node: EntityRef + Debug>(&'a CallGraph<Node>);
+
+        impl<'a, Node: EntityRef + Debug> Debug for Edges<'a, Node> {
+            fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+                f.debug_map()
+                    .entries(self.0.nodes().map(|n| (n, self.0.edges(n))))
+                    .finish()
+            }
+        }
+
+        f.debug_struct("CallGraph")
+            .field("edges", &Edges(self))
+            .finish()
+    }
+}
+
+impl<Node> CallGraph<Node>
+where
+    Node: EntityRef + Debug,
+{
+    /// Construct a new call graph.
+    ///
+    /// `funcs` should be an iterator over all function nodes in this call
+    /// graph's translation unit.
+    ///
+    /// The `get_calls` function should yield (by pushing onto the given `Vec`)
+    /// all of the callee function nodes that the given caller function node
+    /// calls.
+    pub fn new(
+        funcs: impl IntoIterator<Item = Node>,
+        get_calls: impl Fn(Node, &mut Vec<Node>) -> Result<()>,
+    ) -> Result<Self> {
+        let funcs = funcs.into_iter();
+
+        let (min, max) = funcs.size_hint();
+        let capacity = max.unwrap_or_else(|| 2 * min);
+        let mut edges = SecondaryMap::with_capacity(capacity);
+        let mut edge_elems = vec![];
+
+        let mut calls = vec![];
+        for caller in funcs {
+            calls.clear();
+            get_calls(caller, &mut calls)?;
+
+            let start = edge_elems.len();
+            let start = u32::try_from(start).unwrap();
+            edge_elems.extend_from_slice(&calls);
+            let end = edge_elems.len();
+            let end = u32::try_from(end).unwrap();
```

Comment:
Idle thought -- we have this idiom a lot (list-of-lists stored in a flattened manner, with ranges in another list). It might be nice to build an abstraction around it -- something like

```rust
let mut elems = edge_elems.start_range();
elems.extend_from_slice(...); // or elems.push or whatever
edges[caller] = elems.finish();
```

not necessarily in this PR but as a cleanup at some point.

We do have `Ranges` which optimizes for the case that we build the ranges themselves in-order so we can store only one end of the range (and use previous-end as start); here we're processing callers in arbitrary order.

Reply:
(Actually, on reading further down and seeing more instances of this pattern -- I suspect it might be a nice cleanup in this PR!)